### PR TITLE
Replace missing template values with empty string

### PR DESCRIPTION
**Pull request by briandealwis**
Fixes: #6128

**Description**
Introduce new `default` function, modelled on [sprig's `default` function](https://masterminds.github.io/sprig/defaults.html#default):

     {{default "value" .MISSING}}

results in `"value"` if `.MISSING` is nil or a zero value.

### CODE REVIEW

**Review by gsquared94 — DISMISSED** [review]

**Review by nkubala — DISMISSED** [review]
https://github.com/GoogleContainerTools/skaffold/issues/6128#issuecomment-876029104

I'm fine with this, but `ParseEnvTemplate()` is also used when computing tags in the `EnvTagger`. I suppose this could change some behavior that users may have been relying on when their referenced values aren't set....but:

1) that's probably a very fringe corner case, and
2) this is probably the way we should have been doing this to begin with.

**Comment by tejal29** [comment]
Why not `template.New("envTemplate").Option("missingkey=error").Parse(t)` and error out when evaluating the template?

**Comment by briandealwis** [comment]
> Why not template.New("envTemplate").Option("missingkey=error").Parse(t) and error out when evaluating the template?

I think the general expectation is that we should handle environment variables like the shell, which replaces non-existent variables with "".
```
$ echo hi${THIS_DOES_NOT_EXIST}there
hithere
```

**Comment by briandealwis** [comment]
That said, @tejal29's suggestion has some appeal.  We could provide something like Helm's [`default` function](https://helm.sh/docs/chart_template_guide/functions_and_pipelines/#using-the-default-function).  That comes from the [`sprig` template function library](https://github.com/Masterminds/sprig) that Helm incorporates; that [library adds more than a hundred functions](http://masterminds.github.io/sprig/).

**Comment by briandealwis** [comment]
I've reverted our template behaviour back to the original behaviour and introduced a new `default` function modelled on [the sprig project's `default`](https://masterminds.github.io/sprig/defaults.html#default).

File: pkg/skaffold/util/env_template.go [hunk 1]
```diff
@@ -131,3 +135,27 @@ func MapToFlag(m map[string]*string, flag string) ([]string, error) {
...
+
+// defaultFunc is a template function that behaves as sprig's default function.
+// See https://masterminds.github.io/sprig/defaults.html#default
+func defaultFunc(dflt, value interface{}) interface{} {
+	if value == nil {
+		return dflt
+	}
+	v := reflect.ValueOf(value)
+	switch v.Kind() {
+	case reflect.Array, reflect.Slice:
```
Comment (nkubala) [inline pkg/skaffold/util/env_template.go:147]: are there other types we want to consider here? or another way of posing this question: how much of https://github.com/Masterminds/sprig/blob/master/defaults.go#L35-L60 should we reuse here?

**Review by nkubala — COMMENTED** [review]
nice. I wonder how many other functions from https://masterminds.github.io/sprig/ would be useful to support here... 🤔

File: pkg/skaffold/util/env_template.go (hunk 1 above)
Comment (briandealwis) [reply pkg/skaffold/util/env_template.go:147]: Given that most of our uses will be environment strings, I don't expect any.

**Review by nkubala — APPROVED** [review]

**Pull request merged**